plugin/pkg/client/auth/oidc/oidc.go

```diff
@@ -112,146 +159,134 @@ func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.A
+	return cache.setClient(issuer, clientID, clientSecret, provider), nil
 }
 
 type oidcAuthProvider struct {
-	refresher      *idTokenRefresher
-	initialIDToken jose.JWT
+	// Interface rather than a raw *oidc.Client for testing.
+	client OIDCClient
+
+	// Stubbed out for testing.
+	now func() time.Time
+
+	// Mutex guards presisting to the kubeconfig file and allows synchronized
+	// updates to the in-memory config. It also ensures concurrent calls to
+	// the RoundTripper only trigger a single refresh request.
+	mu        sync.Mutex
+	cfg       map[string]string
+	persister restclient.AuthProviderConfigPersister
 }
 
-func (g *oidcAuthProvider) WrapTransport(rt http.RoundTripper) http.RoundTripper {
-	at := &oidc.AuthenticatedTransport{
-		TokenRefresher: g.refresher,
-		RoundTripper:   rt,
-	}
-	at.SetJWT(g.initialIDToken)
+func (p *oidcAuthProvider) WrapTransport(rt http.RoundTripper) http.RoundTripper {
 	return &roundTripper{
-		wrapped:   at,
-		refresher: g.refresher,
+		wrapped:  rt,
+		provider: p,
 	}
 }
 
-func (g *oidcAuthProvider) Login() error {
+func (p *oidcAuthProvider) Login() error {
 	return errors.New("not yet implemented")
 }
 
 type OIDCClient interface {
 	refreshToken(rt string) (oauth2.TokenResponse, error)
-	verifyJWT(jwt jose.JWT) error
+	verifyJWT(jwt *jose.JWT) error
 }
 
 type roundTripper struct {
-	refresher *idTokenRefresher
-	wrapped   *oidc.AuthenticatedTransport
+	provider *oidcAuthProvider
+	wrapped  http.RoundTripper
 }
 
 func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
-	var res *http.Response
-	var err error
-	firstTime := true
-	wait.ExponentialBackoff(backoff, func() (bool, error) {
-		if !firstTime {
-			var jwt jose.JWT
-			jwt, err = r.refresher.Refresh()
-			if err != nil {
-				return true, nil
-			}
-			r.wrapped.SetJWT(jwt)
-		} else {
-			firstTime = false
-		}
-
-		res, err = r.wrapped.RoundTrip(req)
-		if err != nil {
-			return true, nil
-		}
-		if res.StatusCode == http.StatusUnauthorized {
-			return false, nil
-		}
-		return true, nil
-	})
-	return res, err
-}
-
-type idTokenRefresher struct {
-	cfg           map[string]string
-	client        OIDCClient
-	persister     restclient.AuthProviderConfigPersister
-	intialIDToken jose.JWT
-}
-
-func (r *idTokenRefresher) Verify(jwt jose.JWT) error {
-	claims, err := jwt.Claims()
+	token, err := r.provider.idToken()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	now := time.Now()
-	exp, ok, err := claims.TimeClaim("exp")
-	switch {
-	case err != nil:
-		return fmt.Errorf("failed to parse 'exp' claim: %v", err)
-	case !ok:
-		return errors.New("missing required 'exp' claim")
-	case exp.Before(now):
-		return fmt.Errorf("token already expired at: %v", exp)
+	// shallow copy of the struct
+	r2 := new(http.Request)
+	*r2 = *req
+	// deep copy of the Header
+	r2.Header = make(http.Header)
+	for k, s := range req.Header {
+		r2.Header[k] = s
 	}
+	r2.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
```

Comment:
To comply with the RoundTripper interface

> RoundTrip should not modify the request...

https://golang.org/pkg/net/http/#RoundTripper

If we don't copy deepcopy the header then we're adding a bearer toke to the original header, modifying the original request.